File: Pokémon 3D Server/Pokémon 3D Server/Class/Setting/Data List/IPBlackList.vb
Imports System.Net

Public Class IPBlackList
    Public _IPAddress As IPAddress
    Public Property IPAddress As String
        Set(value As String)
            Try
                _IPAddress = Net.IPAddress.Parse(value)
            Catch ex As Exception
                _IPAddress = Nothing
            End Try
        End Set
        Get
            Return _IPAddress.ToString
        End Get
    End Property
    Public Property BanReason As String
    Public Property BanStartTime As Date
    Public Property BanDuration As Integer

    Public ReadOnly Property BanRemainingTime As String
        Get
            Dim BanRemainingTimes As TimeSpan
            Dim BanRemainingText As String = Nothing

            If BanDuration > 0 Then
                BanRemainingTimes = BanStartTime.AddSeconds(BanDuration) - Date.Now
                If BanRemainingTimes.Days = 1 Then
                    BanRemainingText &= BanRemainingTimes.Days & " day "
                ElseIf BanRemainingTimes.Days > 1
                    BanRemainingText &= BanRemainingTimes.Days & " days "
                End If

                If BanRemainingTimes.Hours = 1 AndAlso BanRemainingTimes.Days = 0 Then
                    BanRemainingText &= BanRemainingTimes.Hours & " hour "
                ElseIf BanRemainingTimes.Hours > 1 AndAlso BanRemainingTimes.Days = 0 Then
                    BanRemainingText &= BanRemainingTimes.Hours & " hours "
                End If

                If BanRemainingTimes.Minutes = 1 AndAlso BanRemainingTimes.Days = 0 Then
                    BanRemainingText &= BanRemainingTimes.Minutes & " minute "
                ElseIf BanRemainingTimes.Minutes > 1 AndAlso BanRemainingTimes.Days = 0 Then
                    BanRemainingText &= BanRemainingTimes.Minutes & " minutes "
                End If

                If BanRemainingTimes.Seconds = 1 AndAlso BanRemainingTimes.Days = 0 Then
                    BanRemainingText &= BanRemainingTimes.Seconds & " second "
                ElseIf BanRemainingTimes.Seconds > 1 AndAlso BanRemainingTimes.Days = 0 Then
                    BanRemainingText &= BanRemainingTimes.Seconds & " seconds "
                End If
            Else
                BanRemainingText = "Permanent"
            End If
            Return BanRemainingText
        End Get
    End Property

    Public Sub New(ByVal IPAddress As String, ByVal BanReason As String, ByVal BanStartTime As Date, ByVal BanDuration As Integer)
        Me.IPAddress = IPAddress
        Me.BanReason = BanReason
        Me.BanStartTime = BanStartTime
        Me.BanDuration = BanDuration
    End Sub
End Class

File: Pokemon.3D.Server.Core/Server_Client_Listener/Settings/Data/IPBlackList.cs
using System;
using System.Net;
using Pokemon_3D_Server_Core.Shared.jianmingyong.Modules;

namespace Pokemon_3D_Server_Core.Server_Client_Listener.Settings.Data
{
    /// <summary>
    /// Class containing IPBlackList.
    /// </summary>
    public class IPBlackList : CommonData
    {
        private IPAddress _IPAddress;
        /// <summary>
        /// Get/Set Player IPAddress
        /// </summary>
        public string IPAddress
        {
            get
            {
                return _IPAddress.ToString();
            }
            set
            {
                try
                {
                    _IPAddress = System.Net.IPAddress.Parse(value);
                }
                catch (Exception ex)
                {
                    ex.CatchError();
                }
            }
        }

        /// <summary>
        /// Get/Set IPBlackList StartTime
        /// </summary>
        public DateTime StartTime { get; set; }

        /// <summary>
        /// Get/Set IPBlackList Duration
        /// </summary>
        public int Duration { get; set; }

        /// <summary>
        /// Get IPBlackList Remaining Time
        /// </summary>
        public string RemainingTime
        {
            get
            {
                string ReturnText = null;

                if (Duration > 0)
                {
                    if (StartTime.AddSeconds(Duration) > DateTime.Now)
                    {
                        TimeSpan RemainingTime = StartTime.AddSeconds(Duration) - DateTime.Now;

                        if (RemainingTime.Days > 1)
                        {
                            ReturnText = RemainingTime.Days.ToString() + " Days";
                        }
                        else if (RemainingTime.Days == 1)
                        {
                            ReturnText = "1 Day";
                        }
                        else
                        {
                            if (RemainingTime.Hours == 1)
                            {
                                ReturnText += "1 Hour ";
                            }
                            else if (RemainingTime.Hours > 1)
                            {
                                ReturnText += RemainingTime.Hours.ToString() + " Hour ";
                            }

                            if (RemainingTime.Minutes == 1)
                            {
                                ReturnText += "1 Minute ";
                            }
                            else if (RemainingTime.Minutes > 1)
                            {
                                ReturnText += RemainingTime.Minutes.ToString() + " Minutes ";
                            }
                            else if (RemainingTime.Minutes == 0 && RemainingTime.TotalSeconds > 60)
                            {
                                ReturnText += "0 Minute ";
                            }

                            if (RemainingTime.Seconds == 1)
                            {
                                ReturnText += "1 Second";
                            }
                            else if (RemainingTime.Seconds > 1)
                            {
                                ReturnText += RemainingTime.Seconds.ToString() + " Seconds";
                            }
                            else if (RemainingTime.Seconds == 0)
                            {
                                ReturnText += "0 Second";
                            }
                        }
                    }
                    else
                    {
                        ReturnText = null;
                    }
                }
                else
                {
                    ReturnText = "Permanent";
                }
                return ReturnText;
            }
        }

        /// <summary>
        /// New IPBlackList
        /// </summary>
        /// <param name="IPAddress">Player IPAddress</param>
        /// <param name="Reason">IPBlackList Reason</param>
        /// <param name="StartTime">IPBlackList Start Time</param>
        /// <param name="Duration">IPBlackList Duration</param>
        public IPBlackList(string IPAddress, string Reason, DateTime StartTime, int Duration)
        {
            this.IPAddress = IPAddress;
            this.Reason = Reason;
            this.StartTime = StartTime;
            this.Duration = Duration;
        }
    }
}